[BEAM-619] keep track of local split sources in UnboundedSourceWrapper

mxm opened the pull request:
CC @aljoscha

mxm commented on runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapper.java:
```diff
@@ -324,9 +331,9 @@ public void cancel() {
-      UnboundedSource.UnboundedReader<OutputT> reader = readers.get(i);
+    for (int i = 0; i < localSplitSources.size(); i++) {
+      UnboundedSource<OutputT, CheckpointMarkT> source = localSplitSources.get(i);
+      UnboundedSource.UnboundedReader<OutputT> reader = localReaders.get(i);
```
FYI: This is where the IndexOutOfBoundsException occurred.

aljoscha commented:
Fix looks legit. We should probably add a test, though, since this apparently was not caught by the tests before.

mxm commented:
The existing test doesn't catch the bug because it executes with parallelism and number of splits set to 1.

aljoscha commented:
Then maybe that test should be changed so it would catch the bug. 😅

mxm commented:
Yep, was already working on it. Pleas see the new commit.

aljoscha commented on runners/flink/runner/src/test/java/org/apache/beam/runners/flink/streaming/UnboundedSourceWrapperTest.java:
```diff
@@ -296,13 +312,13 @@ public void close() {
     cfg.setTimeCharacteristic(TimeCharacteristic.EventTime);
 
-    Environment env = new DummyEnvironment("MockTwoInputTask", 1, 0);
+    Environment env = new DummyEnvironment("MockTwoInputTask", numSubTasks, 0);
```
If you set it up like this the source is always only responsible for one split. Maybe it would be good if a source had several splits, i.e. if the number of source splits was higher than the number of subtasks.

mxm replied:
Thanks for pointing out. I'll introduce a second test parameter for the number of splits.

mxm replied:
Sorry, I meant number of tasks.

mxm replied:
I've updated the test case.

mxm commented:
Updated the test case to test multiple readers per source. Also, I've removed one of the test cases because it is now obsolete with the other test being parameterized.
